wallet-next/src/storage.rs

```diff
@@ -80,4 +86,73 @@ impl Storage {
             result[0].bytes.as_ref().unwrap().as_slice(),
         )?)
     }
+
+    pub async fn record_block(
+        &self,
+        scan_result: ScanResult,
+        nct: &mut NoteCommitmentTree,
+    ) -> anyhow::Result<Self> {
+        // Insert all new note records
+        for note_record in scan_result.new_notes {
+            let query = format!(
+                r#"INSERT INTO notes (
+                        note_commitment,
+                        height_spent,
+                        height_created,
+                        diversifier,
+                        amount,
+                        asset_id,
+                        transmission_key,
+                        blinding_factor,
+                        diversifier_index,
+                        nullifier)
+                    VALUES ({:?},NULL,{:?},{:?},{:?},{:?},{:?},{:?},{:?},{:?})"#,
+                note_record.note_commitment.0.to_bytes(),
+                //height_spent is NULL
+                scan_result.height,
+                note_record.note.diversifier().0,
+                note_record.note.amount(),
+                note_record.note.asset_id().to_bytes(),
+                note_record.note.transmission_key().0,
+                note_record.note.note_blinding().to_bytes(),
+                note_record.diversifier_index.0,
+                note_record.nullifier.to_bytes()
+            );
+
+            let _result = &self.pool.execute(&*query).await?;
+        }
+
+        // Update any rows of the table with matching nullifiers to have height_spent
+        for nullifier in scan_result.spent_nullifiers {
+            let height_spent = format!("{:?}", &scan_result.height);
+            let nullifier = format!("{:?}", &nullifier.to_bytes());
+
+            let result = query!(
+                r#"UPDATE notes 
+                SET height_spent = ?
+                WHERE nullifier = ?
+                RETURNING note_commitment"#,
+                height_spent,
+                nullifier
+            )
+            .fetch_all(&self.pool)
+            .await?;
+
+            //Forget spent note commitments from the NCT
+
+            nct.remove_witness(&Commitment::try_from(result[0].note_commitment.as_slice())?);
+        }
+
+        // Update NCT table with current NCT state
+
+        let query = format!(
+            "UPDATE note_commitment_tree
+            SET bytes = {:?}",
+            bincode::serialize(nct)
+        );
```

Comment:
Any reason not to use `query!` here rather than building a string?

Reply:
Either can work, but the string interpolation for `query!` doesn't support inline functions, so, personal preference vs. serializing the `nct` into a variable beforehand, I suppose. `query!` also doesn't infer string formatting for every type, so that may also need to be done outside the macro. No strong opinions either way, I still haven't found a way to use `sqlx` that feels succinct & maximally readable to me.